### ISSUE TITLE
Renovate Bot not updating all packages in group together

### ISSUE THREAD

**Issue by MartinPetkov** [renovatebot/config-help#782]
In our repo, we have a [renovate.json](https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/blob/master/renovate.json) and we recently configured it to group all Terraform package updates together.

**Expectation:** All Terraform modules are updated in the same PR

**Reality:** They are not. See https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/pull/497. It only updated files under the `examples/` directory, but I expected it to also updated files under the `templates/` directory (specifically [this one](https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/blob/master/templates/tfengine/components/resources/bastion_hosts/main.tf#L18)). It could possibly be related to the fact that I closed https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/pull/492 prematurely.

I'm not sure if this is a bug or a misconfiguration. Any help is appreciated.

**Comment by rarkins** [comment]
I think that the package rule may be fine, but the dependency is perhaps not being detected in the first place. When I check the log for extracted packages, that file is not found:

![image](https://user-images.githubusercontent.com/6311784/88755697-076ff780-d162-11ea-9b66-58f6b643500a.png)

This is an indication that Renovate found 0 dependencies when it attempted to parse it. Considering the heavy templating (e.g. `module "{{resourceName . "name"}}"`) it seems like the best guess for now.

BTW normally Renovate would use the `module` contents as the dependency name, which would product a pretty weird looking PR description too.

Which templating engine is in use?

**Comment by viceice** [comment]
@MartinPetkov As it isn't easy to detect the dependency when templating is in use i would suggest to use the regex manager for that.

so prefix the version with a 
```hcl
  # renovate: datasource=terraform-module depName=terraform-google-modules/bastion-host/google 
  version = "~> 2.7.0"
```

and add a regex to detect this

```js
 {
            "fileMatch": [
                "\\.tf$"
            ],
            "matchStrings": [
                "# renovate: datasource=(?<datasource>[a-z-]+?) depName=(?<depName>.+?)(?: versioning=(?<versioning>[a-z-]+?))?\\s+version\\s+=\\s+\"(?<currentValue>.+?)\"\s"
            ],
            "versioningTemplate": "{{#if versioning}}{{versioning}}{{else}}semver{{/if}}"
        }
```

it should also be possible to reuse the `source` line to match the `depName`

**Comment by MartinPetkov** [comment]
Thank you both for the detailed response.

@rarkins The templating is [Go Templating](https://golang.org/pkg/text/template/). Our software is a wrapper around that, focused on templating Terraform configs. Is it possible to use packageFile as the module name? That would be fine for PR descriptions.

@viceice What does "dependency" mean in the context of Renovate Bot? And could I just use the `matchStrings` to match the existing `source = "terraform-google-modules/bastion-host/google"` and `version = "~> 2.7.0"` lines? I would expect Renovate Bot to recognize those even if the module name is a template value.

**Comment by rarkins** [comment]
The short explanation is that Renovate is currently designed to parse valid TF files and the raw templates are not. So the only reasonable expectation for now is that Tf files with invalid syntax won't be detected.

Important to note that we're not ideologically against it - it's just unrealistic in most cases for us to parse invalid syntax, whether it's Tf or JSON. That said, Tf may be achievable because we do very limited parsing anyway. 

The regex manager is designed partly for such situations so what you can identify dependencies in any file type you want and fall back to that if Renovates default parsing can't cut it.

**Comment by MartinPetkov** [comment]
Okay. I'll experiment with the regex manager.

Is there a way to run Renovate locally to see what change it should identify?

**Comment by rarkins** [comment]
Yes you can run the CLI via npm or Docker

**Comment by MartinPetkov** [comment]
Thank you. I'll get back to this issue after some experimentation.

**Comment by rarkins** [comment]
Also in case it wasn't clear;

- you can almost definitely since it yourself by using the regex manager, but
- we're open to suggestions of how to improve the existing terraform parsing if supporting templates can be reasonably done

**Comment by MartinPetkov** [comment]
I'm struggling with the hundreds of flags to the npm CLI. Is there a simple way I can run it locally, and just see what dependencies it identified, without having it open PRs?

**Comment by MartinPetkov** [comment]
I tried, but I can't seem to get it to respect my configuration, or to use a locally-checked out repository. It always uses a repository on GitHub, and it always uses its renovate.json instead of the one I have locally, or the config I have in the `onboardingConfig` block. This is my config.js:

```js
module.exports = {
  endpoint: 'https://api.github.com',
  token: '<redacted>',
  platform: 'github',
  logFileLevel: 'warn',
  logLevel: 'debug',
  logFile: './renovate.log',
  onboarding: true,
  onboardingConfig: {
    "ignorePaths": [],
    "packageRules": [
      {
        "managers": ["terraform"],
        "groupName": "martin",
        "fileMatch": [
          "\\.tf$"
        ],
        "matchStrings": [
          "source\\s+=\\s+\"(?<depName>.+?)\"\\s+version\\s+=\\s+\"(?<currentValue>.+?)\""
        ],
        "datasourceTemplate": "terraform-module",
        "versioningTemplate": "hashicorp"
      }
    ]
  },
  repositories: ['GoogleCloudPlatform/healthcare-data-protection-suite'],
};
```

The logs are also hard to parse (at debug level), I don't really understand where I'm supposed to be looking to see if the regex match worked.

Any help would be appreciated here. I'm really not trying to do anything complicated, I just want to see the regex match in action without committing a string of PRs.

**Comment by rarkins** [comment]
Renovate always runs against a VCS server and never locally. You could try setting enabledManagers to regex only and that way the rest are never run.

**Comment by MartinPetkov** [comment]
I think I have it working against [my fork](https://github.com/MartinPetkov/healthcare-data-protection-suite/blob/master/renovate.json). What should I look for in the debug logs to know that it worked? I see some lines that indicate it might be working, but I'm not sure.

**Comment by rarkins** [comment]
The "packageFiles with updates" log message lists all detected dependencies and their updates

**Comment by MartinPetkov** [comment]
I finally got it working via https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/pull/520. Thanks for the help.
